integration/container/nat_test.go

```diff
@@ -29,7 +29,14 @@ func TestNetworkNat(t *testing.T) {
 	startServerContainer(ctx, t, msg, 8080)
 
 	endpoint := getExternalAddress(t)
+	try := 0
+retry:
 	conn, err := net.Dial("tcp", net.JoinHostPort(endpoint.String(), "8080"))
+	if err != nil && try < 10 {
+		try++
+		time.Sleep(200 * time.Millisecond)
+		goto retry
+	}
```

Comment:
I'm not really a fan of the go labels, maybe we can refactor this to be:

```suggestion
var conn *net.Conn
var err error

for i:=0; i < 10; i++ {
  conn, err = net.Dial("tcp", net.JoinHostPort(endpoint.String(), "8080"))
  if err == nil && conn != nil {
    break
   }
   time.Sleep(200 *time.Milisecond)
  }
  
  assert.NilError(t, err, "could not dial ...")
  assert.NotNil(t, conn, "connection should not be nil")
```

Reply:
Yeah, not a fan generally as well. This was part of the original patch that I took it from. I was considering rewriting it to a `poll.Check` (and to use `poll.WaitOn` here) but chose to keep the changes as they were;
https://github.com/moby/moby/blob/f052dbb54bdfd4dfcb8b007c0dbf33f533ecffd7/vendor/gotest.tools/v3/poll/check.go#L8-L9

I can have a quick try though

Reply:
Rewrote it to use `poll.WaitOn`;

```patch
diff --git a/integration/container/nat_test.go b/integration/container/nat_test.go
index 839619f8e5..5219548fc5 100644
--- a/integration/container/nat_test.go
+++ b/integration/container/nat_test.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"io"
 	"net"
+	"strconv"
 	"strings"
 	"testing"
 	"time"
@@ -25,20 +26,25 @@ func TestNetworkNat(t *testing.T) {
 
 	ctx := setupTest(t)
 
-	msg := "it works"
-	startServerContainer(ctx, t, msg, 8080)
+	const msg = "it works"
+	const port = 8080
+	startServerContainer(ctx, t, msg, port)
 
 	endpoint := getExternalAddress(t)
-	try := 0
-retry:
-	conn, err := net.Dial("tcp", net.JoinHostPort(endpoint.String(), "8080"))
-	if err != nil && try < 10 {
-		try++
-		time.Sleep(200 * time.Millisecond)
-		goto retry
-	}
-	assert.NilError(t, err)
-	defer conn.Close()
+
+	var conn net.Conn
+	addr := net.JoinHostPort(endpoint.String(), strconv.Itoa(port))
+	poll.WaitOn(t, func(t poll.LogT) poll.Result {
+		var err error
+		conn, err = net.Dial("tcp", addr)
+		if err != nil {
+			return poll.Continue("waiting for %s to be accessible: %v", addr, err)
+		}
+		return poll.Success()
+	})
+	defer func() {
+		assert.Check(t, conn.Close())
+	}()
 
 	data, err := io.ReadAll(conn)
 	assert.NilError(t, err)
@@ -47,23 +53,26 @@ retry:
 
 func TestNetworkLocalhostTCPNat(t *testing.T) {
 	skip.If(t, testEnv.IsRemoteDaemon)
-	skip.If(t, testEnv.GitHubActions, "FIXME: https://github.com/moby/moby/issues/41561")
 
 	ctx := setupTest(t)
 
-	msg := "hi yall"
-	startServerContainer(ctx, t, msg, 8081)
-
-	try := 0
-retry:
-	conn, err := net.Dial("tcp", "localhost:8081")
-	if err != nil && try < 10 {
-		try++
-		time.Sleep(200 * time.Millisecond)
-		goto retry
-	}
-	assert.NilError(t, err)
-	defer conn.Close()
+	const msg = "hi yall"
+	const port = 8081
+	startServerContainer(ctx, t, msg, port)
+
+	var conn net.Conn
+	addr := net.JoinHostPort("localhost", strconv.Itoa(port))
+	poll.WaitOn(t, func(t poll.LogT) poll.Result {
+		var err error
+		conn, err = net.Dial("tcp", addr)
+		if err != nil {
+			return poll.Continue("waiting for %s to be accessible: %v", addr, err)
+		}
+		return poll.Success()
+	})
+	defer func() {
+		assert.Check(t, conn.Close())
+	}()
 
 	data, err := io.ReadAll(conn)
 	assert.NilError(t, err)
diff --git a/integration/networking/port_mapping_linux_test.go b/integration/networking/port_mapping_linux_test.go
index 494f4a4ba1..ad87e7cf4a 100644
--- a/integration/networking/port_mapping_linux_test.go
+++ b/integration/networking/port_mapping_linux_test.go
@@ -25,6 +25,7 @@ import (
 	"gotest.tools/v3/assert"
 	is "gotest.tools/v3/assert/cmp"
 	"gotest.tools/v3/icmd"
+	"gotest.tools/v3/poll"
 	"gotest.tools/v3/skip"
 )
 
@@ -254,15 +255,16 @@ func TestProxy4To6(t *testing.T) {
 	inspect := container.Inspect(ctx, t, c, serverId)
 	hostPort := inspect.NetworkSettings.Ports["80/tcp"][0].HostPort
 
-	try := 0
-retry:
-	resp, err := http.Get("http://[::1]:" + hostPort)
-	if err != nil && try < 10 {
-		try++
-		time.Sleep(100 * time.Millisecond)
-		goto retry
-	}
-	assert.NilError(t, err)
+	var resp *http.Response
+	addr := "http://[::1]:" + hostPort
+	poll.WaitOn(t, func(t poll.LogT) poll.Result {
+		var err error
+		resp, err = http.Get(addr)
+		if err != nil {
+			return poll.Continue("waiting for %s to be accessible: %v", addr, err)
+		}
+		return poll.Success()
+	})
 	assert.Check(t, is.Equal(resp.StatusCode, 404))
 }
 
```

Reply:
Failure with this patch looks something like this;


```bash
make BIND_DIR=. TEST_FILTER='TestNetworkLocalhostTCPNat' test-integration
...
=== RUN   TestNetworkLocalhostTCPNat
    nat_test.go:60: timeout hit after 10s: waiting for localhost:8081 to be accessible: dial tcp [::1]:8081: connect: connection refused
--- FAIL: TestNetworkLocalhostTCPNat (10.24s)
```